SPEC.md

```diff
@@ -0,0 +1,195 @@
+   1. A user runs a container with the argument `--device` followed by a device name.
+   2. The container runtime reads the JSON file.
+   3. The container runtime validates that the device is described in the JSON file.
+   4. The container runtime pulls the container image.
+   5. The container runtime generates an OCI specification.
+   6. The container runtime transforms the OCI specification according to the instructions in the JSON file
+
+
+[cni]: https://github.com/containernetworking/cni
+[oci]: https://github.com/opencontainers/runtime-spec
+
+## General considerations
+
+- The device configuration is in JSON format and can easily be stored in a file.
+- The device configuration includes mandatory fields such as "name" and "version".
+- Fields in CDI structures are required unless specifically marked optional.
+
+For the purposes of this proposal, we define two terms very specifically:
+- _container_ can be considered synonymous with a [Linux _network namespace_][namespaces].
+  What unit this corresponds to depends on a particular container runtime implementation. 
+- _device_ which will refer to an actual hardware device.
+
+The key words "must", "must not", "required", "shall", "shall not", "should", "should not", "recommended", "may" and "optonal" are used as specified in [RFC 2119][rfc-2119].
+
+[rfc-2119]: https://www.ietf.org/rfc/rfc2119.txt
+[namespaces]: http://man7.org/linux/man-pages/man7/namespaces.7.html
+
+## CDI JSON Specification
+
+### JSON Definition
+
+```
+{
+    "cdiVersion": "0.1.0",
+    "kind": "<name>",
+    "container-runtime": ["<container-runtime-name>"], (optional)
+
+    "cdiDevices": [
+        {
+            "name": "<name>",
+            "nameShort": ["<short-name>", "<short-name>"], (optional)
+
+            // Same as the below containerSpec field.
+            // This field should only be applied to the Container's OCI spec
+            // if that specific device is requested.
+            "containerSpec": { ... }
+        }
+    ],
+
+    // This field should be applied to the Container's OCI spec if any of the
+    // devices defined above are requested on the CLI
+    "containerSpec": [
+        {
+            "devices": [ (optional)
+                {
+                    "hostPath": "<path>",
+                    "containerPath": "<path>",
+
+                    // Cgroups permissions of the device, candidates are one or more of
+                    // * r - allows container to read from the specified device.
+                    // * w - allows container to write to the specified device.
+                    // * m - allows container to create device files that do not yet exist.
+                    "permissions": "<permissions>" (optional)
+                }
+            ]
+            "mounts": [ (optional)
+                {
+                    "hostPath": "<source>",
+                    "containerPath": "<destination>",
+                    "options": "<OCI Mount Options>", (optional)
```

Comment:
Do we need permission on the mounted files? I do not think that most of the files need to be written or modified.